lib/model/autocomplete.dart

```diff
@@ -251,11 +253,95 @@ class MentionAutocompleteView extends ChangeNotifier {
     notifyListeners();
   }
 
+  List<User>? _sortedUsers;
+
+  int compareByRelevance({
+    required User userA,
+    required User userB,
+    required int? streamId,
+    required String? topic,
+  }) {
+    // TODO(#234): give preference to "all", "everyone" or "stream".
+
+    // TODO(#618): give preference to subscribed users first.
+
+    if (streamId != null) {
+      final conversationPrecedence = store.recentSenders.compareByRecency(
+        userA: userA,
+        userB: userB,
+        streamId: streamId,
+        topic: topic);
+      if (conversationPrecedence != 0) {
+        return conversationPrecedence;
+      }
+    }
+
+    final dmPrecedence = store.recentDmConversationsView.compareByDms(userA, userB);
+    if (dmPrecedence != 0) {
+      return dmPrecedence;
+    }
+
+    if (!userA.isBot && userB.isBot) {
+      return -1;
+    } else if (userA.isBot && !userB.isBot) {
+      return 1;
+    }
+
+    final userAName = store.autocompleteViewManager.autocompleteDataCache
+      .nameForUser(userA);
+    final userBName = store.autocompleteViewManager.autocompleteDataCache
+      .nameForUser(userB);
+    return userAName.compareTo(userBName);
+  }
+
+  List<User> sortByRelevance({
+    required List<User> users,
+    required Narrow narrow,
+  }) {
+    switch (narrow) {
+      case StreamNarrow():
+        users.sort((userA, userB) => compareByRelevance(
+          userA: userA,
+          userB: userB,
+          streamId: narrow.streamId,
+          topic: null));
+      case TopicNarrow():
+        users.sort((userA, userB) => compareByRelevance(
+          userA: userA,
+          userB: userB,
+          streamId: narrow.streamId,
+          topic: narrow.topic));
+      case DmNarrow():
+        users.sort((userA, userB) => compareByRelevance(
+          userA: userA,
+          userB: userB,
+          streamId: null,
+          topic: null));
+      case AllMessagesNarrow():
+        // do nothing in this case for now
+    }
+    return users;
+  }
+
+  void _sortUsers() {
+    final users = store.users.values.toList();
+    _sortedUsers = sortByRelevance(users: users, narrow: narrow);
+  }
+
+  bool _areUsersModified = false;
+  void _usersModified() {
+    _areUsersModified = true;
+  }
+
   Future<List<MentionAutocompleteResult>?> _computeResults(MentionAutocompleteQuery query) async {
     final List<MentionAutocompleteResult> results = [];
-    final Iterable<User> users = store.users.values;
 
-    final iterator = users.iterator;
+    if (_sortedUsers == null) {
+      _sortUsers();
+    }
+
+    final iterator = _sortedUsers!.iterator;
+    store.users.addListener(_usersModified);
```

Comment:
Hmm interesting. Can you elaborate on what this is about?

This would also be good to pull out as its own prep commit — definitely the `store.dart` changes to support it, and perhaps this `addListener` and the associated bits of logic in this file would also go in that commit.

One reason a separate prep commit is helpful is that its commit message provides a natural venue to give the explanation about why it's useful. :slightly_smiling_face:

Reply:
… ah I see, you made a comment just now about this. Reading now…